src/Emulator.ts

```diff
@@ -863,6 +872,57 @@ export class Emulator implements ITxRunnerProvider, IGetGenesisInfos, IGetProtoc
         this.debug(2, `Transaction ${txHash} is valid`);
         return true;
     }
+
+    /**
+     * Slash collateral for a failed transaction
+     * @param tx The transaction that failed phase-2 validation
+     */
+    private slashCollateral(tx: Tx): void {
+        const collateralInputs = tx.body.collateralInputs || [];
+        const collateralPercentage = this.protocolParameters.collateralPercentage || 150; // Default to 150% if not defined
+
+        for (const input of collateralInputs) {
+            const utxoRef = forceTxOutRefStr(input);
+            const utxo = this.utxos.get(utxoRef);
+
+            if (!utxo) {
+                this.debug(0, `Collateral UTxO ${utxoRef} not found in the ledger.`);
+                continue;
+            }
+
+            const fee = tx.body.fee || BigInt(0);
+            const collateralToSlash = (fee * BigInt(collateralPercentage)) / BigInt(100);
+
+            if (utxo.resolved.value.lovelaces > collateralToSlash) {
```

Comment:
```suggestion
            if (utxo.resolved.value.lovelaces >= collateralToSlash) {
```
Why not consider the equal case?

Reply:
And if it's 0, you handle the case of not sending anything to the changeAddress afterwards